### PR TITLE
New bclconverter script

### PR DESCRIPTION
**Pull request by sylvinite**
Note: Paths are hardcoded to local atm since the rsync isnt running.

Fetches some info from the workflow process
Assigns a few thresholds to the process (workflow step)
Reformats laneBarcode.html for usage of other applications
Assigns some info from laneBarcode.html to individual samples in the process
Flags samples as QC PASS/FAIL based on thresholds

Especially input on logging and exceptions would be much appreciated.

### CODE REVIEW

**Comment by sylvinite** [comment]
@Galithil @vezzi

File: scripts/bcl_converter.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
+"""
+This file together with bcl_thresholds.py performs the bclconversion step of LIMS workflow.
+In common tongue, it:
+ 
+Fetches some info from the workflow process
```
Comment (vezzi) [inline scripts/bcl_converter.py]: state clearly the info that is fetched

File: scripts/bcl_converter.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+            elif 'Illumina Sequencing (HiSeq X) 1.0' in k:
+                proc_stats['Run type'] ='HiSeqX10'
+                proc_stats['Data type'] = 'HiSeqX'
+            else:
+                sys.exit("Unhandled prior workflow step (run type)")
+            logging.info("Run type set to " + proc_stats['Run type'])
+            break
+        
+    proc_stats['paired'] = False
+    if 'Read 2 Cycles ' in proc_stats:
```
Comment (vezzi) [inline scripts/bcl_converter.py]: white space?

Comment (vezzi) [inline scripts/bcl_converter.py]: is 'read 2 Cycles' present also in HiSeqX workflow?

File: scripts/bcl_converter.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        proc_stats['paired'] = True
+    logging.info("Paired libraries: " + str(proc_stats['paired']))  
+    
+    return proc_stats
+
+"""Sets run thresholds"""
+#Chemistry for miseq and x10 needs to be expanded
+def manipulate_process(demux_process, proc_stats):      
+    thresholds = Thresholds('HiSeq', proc_stats['Run type'], proc_stats['paired'])
```
Comment (vezzi) [inline scripts/bcl_converter.py]: 'HiSeq' likely to be 'proc_stats['Run type']' and the other data Type

File: scripts/bcl_thresholds.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        
+        #Check only valid values are entered
+        if not instrument in self.valid_instruments or not chemistry in self.valid_chemistry:
+            sys.exit("Detected instrument, chemistry and/or read_setup are not present in bcl_thresholds.py")
+        else:
+            self.set_Q30(instrument, chemistry, paired)
+            self.set_exp_lane_clust(instrument, chemistry, paired)
+    
+    #Q30 values here go for max read length
+    #If read length should be a factor, values need to be upd.
```
Comment (vezzi) [inline scripts/bcl_thresholds.py]: insert comment about the fact that threshold reflect document 1262 and 1244 in AM-System (therefore integrate also read -lenght checks)

File: scripts/bcl_thresholds.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+                    self.Q30 = 60
+        elif instrument == 'HiSeq':
+            if chemistry in ['HiSeq Rapid Flow Cell v1','HiSeq Rapid Flow Cell v2', "TruSeq Rapid Flow Cell v2", "TruSeq Rapid Flow Cell v1"] :
+                self.Q30 = 75
+            elif chemistry == 'HiSeq Flow Cell v3':
+                self.Q30 = 80
+            elif chemistry == 'HiSeq Flow Cell v4':
+                self.Q30 = 80
+        elif instrument == 'HiSeqX10':
+            self.Q30 = 75
```
Comment (vezzi) [inline scripts/bcl_thresholds.py]: add comment that we are running only HiSeqX 2.5 chemistry

File: scripts/bcl_thresholds.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    
+    def set_exp_lane_clust(self, instrument, chemistry, paired):
```
Comment (vezzi) [inline scripts/bcl_thresholds.py]: numbers to be taken from doc 1244

File: scripts/bcl_converter.py [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+                target_file.put()
+            except:
+                sys.exit("Failed to apply artifact data to LIMS")
+    
+
+"""Creates demux_{FCID}_{time}.csv and attaches it to process"""
+#Undetermined are written differently. Might be worth discussing
+#change TEMPORARY LINE once done
+def write_demuxfile(proc_stats):
+    prefix = '/srv/illumina/'+ proc_stats['Data type'] + '_data/nosync/'
```
Comment (vezzi) [inline scripts/bcl_converter.py]: use proper python merger .... os.path.join(path, *paths)

File: scripts/bcl_converter.py [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+    try:
+        appendix = proc_stats['Run ID'] + '/Demultiplexing/Reports/html/' + proc_stats['Flow Cell ID'] + '/all/all/all/laneBarcode.html'
+    except:
+        sys.exit("Unable to set demux filename. Udf does not contain keys for run id and/or flowcell id.")
+    lanebc_path = prefix + appendix
+    #TEMPORARY LINE
+    lanebc_path = '/Users/isaksylvin/SciLifeLab/preprocExampleData/24-142862/laneBarcode.html'
```
Comment (vezzi) [inline scripts/bcl_converter.py]: test code

**Comment by Galithil** [comment]
@sylvinite Anything new on this, or is it stable enough so we can merge ?

**Comment by sylvinite** [comment]
1) Path is still not referring to /srv/mfs (since none is up).
2) Some weird sample values are coming up; have to check flowcell_parser
3) I haven't listed the two dependencies I need (flowcell_parser & click)

**Comment by Galithil** [comment]
As advertised.

**Comment by ewels** [comment]
So what was the conclusions @sylvinite? Keep the dependency but don't specify it in `setup.py`?

**Comment by Galithil** [comment]
"You'll add that dependency to this repo over my dead body"
or something like that

**Pull request merged**